### ISSUE TITLE
Option: Don't Install Default Match All Rule

### ISSUE THREAD

**Issue by craig-riecke** [frenetic-lang/frenetic#429]
Dell switches have some quirky behavior.  If there's a Match All rule in the ACL table - whether it's Match All -> Drop or Match All -> Controller - none of the L2 or L3 rules are fired.  That's because ACL is the default table, and all rules take precedence.  

The Frenetic compiler adds a Match All -> Drop.  If there's a Match All rule in the policies, the Match All -> Drop is thrown out by the optimizer.  If there isn't, it stays in.  The upshot is there's always a Match All rule in the flow tables.  

It turns out the Dell Switches follow the OpenFlow 1.0 standard and send all unmatched packets to the controller.  In the Dell CLI, you can override this and make the unmatched packets drop, but there's no way to set this by the OpenFlow configuration protocol, and it's non-standard anyway.  All of the Gates Dell switches follow the default.  So if we simply remove the Match All -> Drop rule (maybe via an option), the L2 table will work.

**Comment by craig-riecke** [comment]
About halfway done.  It turns out Match All -> Drop Rules are added in two places:

* One is from the FDD -> Flow Table generator.  The default left-most leaf in an FDD is "False", which is technically a node with no predicate and no actions.  OpenFlow interprets a node with no predicate and no actions as "Match All -> Drop.", and that's in fact the rule that's generated.  But it really should be no rule at all.  I committed a change which emits no rule.  Programs that have an explicit Match All rule are not affected.  

* The other is in the DynGraph composer.  When you have more than one app, the policies are are all Unioned together before emitting.  A Match All -> Drop element is used because it's an identity element under Union and it makes writing fold_left's easier.  It's used in a few places.  But it has the effect of creating Drop instructions in places it doesn't belong.  For instance, in the simple program:

if switch = 1 and port =2 then
   Output(Controller)
else if switch = 1 and port = 1 and ethDst = ff:ff:ff:ff:ff:ff then
   Output(Physical(2))

The table:

Inport = 2, Action: Output(Controller)
Inport = 1, EthDst = ff:ff:ff:ff:ff:ff, Action: Output(Physical(2))
Inport = 1, Drop

is produced.  The last rule was not explicitly asked for, and I would argue it's incorrect.  But this only happens when you use apps.  If you use Frenetic_NetKAT_Local_Compiler directly on that policy (as Frenetic Shell does), the drop rule doesn't appear.

**Comment by craig-riecke** [comment]
Bluggh.  #2 is actually more complicated.  When NetKAT makes FDD's, it creates False branches for lots of different situations, not just Unions.  A simple NetKAT program like:

```
Filter( Inport = 1 and EthDst = ff:ff:ff:ff:ff:ff) ; Output(Physical(2))
```

will generate a flow table:

```
Inport = 1, EthDst = ff:ff:ff:ff:ff:ff, Action: Output(Physical(2))
Inport = 1, Drop
```

In a learning switch, you might have additional rules to forward unicast packets with a  certain Mac address to their next hop.  These will work fine on Mininet and most switches, but not on a Dell.   

It'd be nice if we could just generate EthDst = ff:ff:...:ff with an action NO-OP.  But in OpenFlow there is no NO-OP.  And the intuitive representation of NO-OP, a rule with no actions, in OpenFlow means "drop the packet".  

I've hacked Frenetic to simply not output any rules with no actions.  This will probably make the Dell switches work properly, but it's bad in general because (1) there's no way to explicitly drop a packet (2) it's unnecessary for most switches out there.  

I can't think of a better solution at the moment.

**Comment by craig-riecke** [comment]
Blurrgh.  Now that I've taken out the explicit Match All rules, there's another problem.  Packets that generate a table-miss, which sends a Packet In message to Frenetic, are dropped there.   There's some multiplexing going on with Output(Pipe('my_pipe_name')) that isn't happening on a table miss ... if someone can enlighten me on this, it'd save me from digging through the code...

**Comment by jnfoster** [comment]
I can do my best to enlighten... or at least explain what's going on :-)

Chat in 432?

-N

**Comment by craig-riecke** [comment]
The proposed solution was a chainsaw.  Nate came up with a scalpel-solution, implemented in Pull Request #416 as "remove_tail_drops."  It's off-by-default, but does the job when turned on for Dell switches.
